src/models/events.rs

```diff
@@ -0,0 +1,294 @@
+#[non_exhaustive]
+pub enum EventType {
+    PushEvent,
+    CreateEvent,
+    UnknownEvent(String),
+}
+
+/// The repository an [`Event`] belongs to.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[non_exhaustive]
+pub struct Repository {
+    pub id: u64,
+    pub name: String,
+    pub url: Url,
+}
+
+/// The organization an [`Event`] belongs to.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[non_exhaustive]
+pub struct Org {
+    pub id: u64,
+    pub login: String,
+    pub gravatar_id: String,
+    pub url: Url,
+    pub avatar_url: Url,
+}
+
+/// The actor that created this [`Event`].
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[non_exhaustive]
+pub struct Actor {
+    pub id: u64,
+    pub login: String,
+    pub display_login: String,
+    pub gravatar_id: String,
+    pub url: Url,
+    pub avatar_url: Url,
+}
+
+impl<'de> Deserialize<'de> for Event {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct Intermediate {
+            id: String,
+            #[serde(rename = "type")]
+            typ: String,
+            actor: Actor,
+            repo: Repository,
+            public: bool,
+            created_at: DateTime<Utc>,
+            org: Option<Org>,
+            payload: Option<serde_json::Value>,
+        }
+        let intermediate = Intermediate::deserialize(deserializer)?;
+        let event_type = deserialize_event_type(intermediate.typ.as_ref());
+        let payload = intermediate.payload.map_or(Ok(None), |data| {
+            deserialize_payload(&event_type, data).map_err(|e| Error::custom(e.to_string()))
+        })?;
+        let event = Event {
+            id: intermediate.id,
+            typ: event_type,
+            actor: intermediate.actor,
+            repo: intermediate.repo,
+            public: intermediate.public,
+            created_at: intermediate.created_at,
+            org: intermediate.org,
+            payload,
+        };
+        Ok(event)
+    }
+}
+
+fn deserialize_event_type(event_type: &str) -> EventType {
+    match event_type {
+        "CreateEvent" => EventType::CreateEvent,
+        "PushEvent" => EventType::PushEvent,
+        unknown => EventType::UnknownEvent(unknown.to_owned()),
+    }
+}
+
+fn deserialize_payload(
+    event_type: &EventType,
+    data: serde_json::Value,
+) -> Result<Option<EventPayload>, serde_json::Error> {
+    let maybe_payload = match event_type {
+        EventType::PushEvent => serde_json::from_value::<PushEventPayload>(data)
+            .map(|payload| Some(EventPayload::PushEvent(payload)))?,
+        EventType::CreateEvent => serde_json::from_value::<CreateEventPayload>(data)
+            .map(|payload| Some(EventPayload::CreateEvent(payload)))?,
+        _ => serde_json::from_value::<serde_json::Value>(data)
+            .map(|payload| Some(EventPayload::UnknownEvent(payload)))?,
+    };
```

Comment:
I believe this match statement can be shortened to the following, though you should test it.
```suggestion
    let maybe_payload = Some(match event_type {
        EventType::PushEvent => serde_json::from_value(data)
            .map(EventPayload::PushEvent)?,
        EventType::CreateEvent => serde_json::from_value(data)
            .map(EventPayload::CreateEvent)?,
        _ => EventPayload::UnknownEvent(data),
    });
```

Reply:
Yep much cleaner. Not sure how I overlooked the fact `data` is already a `serde_json::Value`, in the unknown case :smile: